File: C++/122_Best Time to Buy and Sell Stock II.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int pos = (int)prices.size() - 1;
        int profit = 0;
        while (pos > 0) {
            int sp = findSell(prices, pos);
            if (sp <= 0)
                break;
            int bp = findBuy(prices, sp);
            profit += prices[sp] - prices[bp];
            pos = bp - 1;
        }
        return profit;
    }
private:
    int findSell(const vector<int> &prices, int pos) {
        while (pos > 0 && prices[pos] <= prices[pos - 1])
            pos--;
        return pos;
    }
    int findBuy(const vector<int> &prices, int pos) {
        while (pos > 0 && prices[pos] > prices[pos - 1])
            pos--;
        return pos;
    }
};

File: Python/122_Best Time to Buy and Sell Stock II.py
class Solution:
    # @param {integer[]} prices
    # @return {integer}
    def maxProfit(self, prices):
        if not prices or len(prices) == 1:
            return 0
        N = len(prices)
        i = 0
        res = 0
        while i < N:
            j = i + 1
            while j < N and prices[j] >= prices[j - 1]:
                j += 1
            if j > i + 1:
                res += prices[j - 1] - prices[i]
            i = j
        return res